File: CardViewObjC/CardViewObjC/BottomSheet/BottomSheetViewController.m
//
//  BottomSheetViewController.m
//  navi-lite
//
//  Created by Phineas.Huang on 2020/5/29.
//  Copyright © 2020 Garmin. All rights reserved.
//

#import "BottomSheetViewController.h"

typedef enum Director {
    up,
    down,
}Director;

typedef enum State {
    partial,
    expanded,
    full
}State;

@interface BottomSheetViewController ()

@property (assign, nonatomic) State lastStatus;

@end

@implementation BottomSheetViewController

static CGFloat fullViewYPosition = 0;
static CGFloat partialViewYPosition = 0;
static CGFloat expandedViewYPosition = 0;

#pragma mark -

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    [self setupData];
    [self setupGestureEvent];

}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];

    __weak __typeof(self)weakSelf = self;
    [UIView animateWithDuration:0.6 animations:^{
        [weakSelf moveView:weakSelf.lastStatus];
    }];
}

#pragma mark -

- (void)setupData {
    fullViewYPosition = 100;
    partialViewYPosition = [UIScreen mainScreen].bounds.size.height - 130;
    expandedViewYPosition = [UIScreen mainScreen].bounds.size.height - 130 - 130;

    self.lastStatus = expanded;
}

#pragma mark -

- (void)setupGestureEvent {
    UIPanGestureRecognizer *gesture = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(panGesture:)];
    [self.view addGestureRecognizer:gesture];
    [self roundViews];
}

- (void)moveView:(State)state {
    CGFloat yPosition = fullViewYPosition;
    if (state == partial) {
        yPosition = partialViewYPosition;

    } else if (state == expanded) {
        yPosition = expandedViewYPosition;
    }

    const CGFloat width = self.view.frame.size.width;
    const CGFloat height = self.view.frame.size.height;
    CGRect rect = CGRectMake(0 , yPosition, width, height);

    self.view.frame = rect;
}

- (void)moveViewWithGesture:(UIPanGestureRecognizer *)recognizer {
    const CGPoint translation = [recognizer translationInView:self.view];
    const CGFloat minY = self.view.frame.origin.y;

    if ((minY + translation.y >= fullViewYPosition) &&
        (minY + translation.y <= partialViewYPosition)) {
        const CGFloat width = self.view.frame.size.width;
        const CGFloat height = self.view.frame.size.height;
        const CGRect rect = CGRectMake(0 , minY + translation.y, width, height);
        self.view.frame = rect;
        [recognizer setTranslation:CGPointZero inView:self.view];
    }
}

- (void)roundViews {
    self.view.layer.cornerRadius = 10;
    self.view.clipsToBounds = YES;
}

- (void)panGesture:(UIPanGestureRecognizer *)recognizer {
    [self moveViewWithGesture:recognizer];

    if (recognizer.state != UIGestureRecognizerStateEnded) {
        return;
    }

    __weak __typeof(self)weakSelf = self;
    [UIView animateKeyframesWithDuration:0.2
                                   delay:0.0
                                 options:UIViewKeyframeAnimationOptionAllowUserInteraction
                              animations:^{
        const Director director = [recognizer velocityInView:weakSelf.view].y >= 0 ? down: up;
        State state = weakSelf.lastStatus;

        if (weakSelf.lastStatus == partial && director == up) {
            state = expanded;

        } else if (weakSelf.lastStatus == expanded && director == up) {
            state = full;
        }

        if (weakSelf.lastStatus == full && director == down) {
            state = expanded;

        } else if (weakSelf.lastStatus == expanded && director == down) {
            state = partial;
        }

        // handle
        if (recognizer.view) {
            if (state == expanded) {
                CGFloat endLocation = recognizer.view.frame.origin.y;
                if (endLocation > expandedViewYPosition &&
                    director == down) {
                    state = partial;

                } else if (endLocation < expandedViewYPosition &&
                           director == up) {
                    state = full;
                }

            } else if (state == partial &&
                weakSelf.lastStatus == partial) {
                CGFloat endLocation = recognizer.view.frame.origin.y;
                if (endLocation < expandedViewYPosition) {
                    state = expanded;
                }
            }
        }

        weakSelf.lastStatus = state;
        [weakSelf moveView:state];
                              } completion:^(BOOL finished) {
                                  // block fires when animaiton has finished
    }];
}

#pragma mark -

@end

File: CardViewSwift/CardViewSwift/BottomSheet/BottomSheetViewController.swift
//
//  BottomSheetViewController.swift
//  navi-lite
//
//  Created by Phineas.Huang on 2020/5/22.
//  Copyright © 2020 Garmin. All rights reserved.
//

import UIKit

extension BottomSheetViewController {
    private enum Director {
        case up
        case down
    }

    private enum State {
        case partial
        case expanded
        case full
    }

    private enum Constant {
        static let fullViewYPosition: CGFloat = 100
        static var partialViewYPosition: CGFloat { UIScreen.main.bounds.height - 130 }
        static var expandedViewYPosition: CGFloat { UIScreen.main.bounds.height - 130 - 130 }
    }
}

class BottomSheetViewController: UIViewController {
    private var lastStatus:State = .partial

    override func viewDidLoad() {
        super.viewDidLoad()

        let gesture = UIPanGestureRecognizer.init(target: self, action: #selector(panGesture))
        view.addGestureRecognizer(gesture)
        roundViews()
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        UIView.animate(withDuration: 0.6, animations: {
            self.moveView(state: .expanded)
        })
    }

    private func moveView(state: State) {
        var yPosition = Constant.fullViewYPosition
        if (state == .partial) {
            yPosition = Constant.partialViewYPosition

        } else if (state == .expanded) {
            yPosition = Constant.expandedViewYPosition
        }

        view.frame = CGRect(x: 0, y: yPosition, width: view.frame.width, height: view.frame.height)
    }

    private func moveView(panGestureRecognizer recognizer: UIPanGestureRecognizer) {
        let translation = recognizer.translation(in: view)
        let minY = view.frame.minY

        if (minY + translation.y >= Constant.fullViewYPosition) && (minY + translation.y <= Constant.partialViewYPosition) {
            view.frame = CGRect(x: 0, y: minY + translation.y, width: view.frame.width, height: view.frame.height)
            recognizer.setTranslation(CGPoint.zero, in: view)
        }
    }

    @objc private func panGesture(_ recognizer: UIPanGestureRecognizer) {
        moveView(panGestureRecognizer: recognizer)

        if recognizer.state == .ended {
            UIView.animate(withDuration: 0.3, delay: 0.0, options: [.allowUserInteraction], animations: {

                // handle velocity
                let director: Director = recognizer.velocity(in: self.view).y >= 0 ? .down : .up
                var state:State = self.lastStatus
                if (self.lastStatus == .partial && director == .up) {
                    state = .expanded

                } else if (self.lastStatus == .expanded && director == .up) {
                    state = .full
                }

                if (self.lastStatus == .full && director == .down) {
                    state = .expanded

                } else if (self.lastStatus == .expanded && director == .down) {
                    state = .partial
                }

                // handle
                if (state == .expanded) {
                    if let endLocation = recognizer.view?.frame.origin.y {
                        if (endLocation > Constant.expandedViewYPosition &&
                            director == .down) {
                            state = .partial

                        } else if (endLocation < Constant.expandedViewYPosition &&
                            director == .up) {
                            state = .full
                        }
                    }

                } else if (state == .partial &&
                    self.lastStatus == .partial) {
                    if let endLocation = recognizer.view?.frame.origin.y {
                        if (endLocation < Constant.expandedViewYPosition) {
                            state = .expanded
                        }
                    }
                }

                self.lastStatus = state
                self.moveView(state: state)
            }, completion: nil)
        }
    }

    func roundViews() {
        view.layer.cornerRadius = 10
        view.clipsToBounds = true
    }
}
